src/components/mc/cuda/mc_cuda.c

```diff
@@ -212,50 +224,172 @@ static ucc_status_t ucc_mc_cuda_init()
     }
 
     ucc_mc_cuda.task_strm_type = cfg->task_strm_type;
+    // spinlock assures 1 thread only initiates mpool and allocates its buffers
+    ucc_spinlock_init(&ucc_mc_cuda.mpool_init_spinlock, 0);
 
     return UCC_OK;
 }
 
 static ucc_status_t ucc_mc_cuda_finalize()
 {
     CUDACHECK(cudaStreamDestroy(ucc_mc_cuda.stream));
+    if (ucc_mc_cuda.mpool_init_flag) {
+        ucc_mpool_cleanup(&ucc_mc_cuda.mpool, 1);
+        ucc_mc_cuda.mpool_init_flag     = 0;
+        ucc_mc_cuda.super.ops.mem_alloc = ucc_mc_cuda_mem_alloc_pool_with_init;
+    }
+    ucc_spinlock_destroy(&ucc_mc_cuda.mpool_init_spinlock);
     return UCC_OK;
 }
 
-static ucc_status_t ucc_mc_cuda_mem_alloc(void **ptr, size_t size)
+static ucc_status_t
+ucc_mc_cuda_mem_alloc(ucc_mc_buffer_header_t **ptr, size_t size)
 {
-    cudaError_t st;
-
-    st = cudaMalloc(ptr, size);
+    cudaError_t             st;
+    ucc_mc_buffer_header_t *h =
+        ucc_malloc(sizeof(ucc_mc_buffer_header_t), "mc cuda");
+    if (ucc_unlikely(!h)) {
+        mc_error(&ucc_mc_cuda.super, "failed to allocate %zd bytes",
+                 sizeof(ucc_mc_buffer_header_t));
+        return UCC_ERR_NO_MEMORY;
+    }
+    st = cudaMalloc(&h->addr, size);
     if (ucc_unlikely(st != cudaSuccess)) {
         cudaGetLastError();
         mc_error(&ucc_mc_cuda.super,
                  "failed to allocate %zd bytes, "
                  "cuda error %d(%s)",
                  size, st, cudaGetErrorString(st));
+        ucc_free(h);
+        return UCC_ERR_NO_MEMORY;
+    }
+    mc_debug(
+        &ucc_mc_cuda.super, "ucc_mc_cuda_mem_alloc size:%zd",
+        size); // TODO: should i keep this? add it also to chunk alloc? in cpu?
```

Comment:
i think this debug is not necessary - when debugging we can always assume that if we don't see error messages then allocation suceeded (and we know the size and number of elems from config).